### PR TITLE
New function parallel::GridTools::exchange_local_bounding_boxes

### PR DESCRIPTION
**Pull request by GivAlz**
Hi everyone!
  I shall shortly update the PR https://github.com/dealii/dealii/pull/5117 hoping it shall be finally approved... Meanwhile I'm working on this function which exchanges the bounding boxes among all processes (and made a test): to exchange the data I transform everything back and forth into a vector of doubles.

I know there are ways to create MPI types...but I've seen this approach many times: it was easier to program and, I guess, sending double is lighter than sending classes around. Anyway suggestions are well accepted!

Best,
Giovanni

### CODE REVIEW

**Comment by tjhei** [comment]
> I know there are ways to create MPI types...but I've seen this approach many times: it was easier to program and, I guess, sending double is lighter than sending classes around. Anyway suggestions are well accepted!

I know that this is not a pretty way to do it, but I would probably just transfer sizeof(BoundingBox) MPI_CHARs or 
```
static_assert(sizeof(BoundingBox)==count*sizeof(double));
MPI_Datatype bbtype;
MPI_Type_contiguous(count,MPI_DOUBLE,&bbtype);
MPI_Type_commit(&bbtype);
```

Then you can skip copying values in/out of the bounding box, which should make the code simpler.

**Comment by bangerth** [comment]
I think copying back and forth does not cost anything compared to the transfer of data, and it avoids all of the vagaries of using bitwise transfer in favor of using the correct data types. (I know that in the old days, MPI was thought of being able to connect machines that have different representations of numbers, whereas today it is in practice only used for machines that all have the same data representation on all participating processes. But I'm still partial towards using the correct data types in transfers :-) )

**Comment by GivAlz** [comment]
Ok, so this is what I call a dilemma :) Thank you Timo for showing how to send it using MPI_Types!
I talked with @luca-heltai also about this: he actually would use a third approach...but, since the current version works decently and this is a matter of how to "approach technically" the data transfer and nothing else, I guess we can leave it as it is (if there are no other problems around) and leave technicalities to whoever wants to cope with them ;)

**Comment by tjhei** [comment]
> I guess we can leave it as it is

I am fine with how it works right now. I just wanted to point out the alternatives.

**Comment by GivAlz** [comment]
> I am fine with how it works right now. I just wanted to point out the alternatives.

I know, next time I might avoid packing/unpacking the data manually, so thanks :)

File: include/deal.II/distributed/grid_tools.h [hunk 1]
```diff
@@ -130,6 +131,19 @@ namespace parallel
     exchange_cell_data_to_ghosts (const MeshType &mesh,
                                   const std::function<boost::optional<DataType> (const typename MeshType::active_cell_iterator &)> &pack,
                                   const std::function<void (const typename MeshType::active_cell_iterator &, const DataType &)> &unpack);
+
+    /**
+     * Exchange with all processors of the Mpi Comm @p mpi_communicator the vector of bounding
```
Comment (davydden) [inline include/deal.II/distributed/grid_tools.h]: Mpi Comm -> MPI communicator

File: include/deal.II/distributed/grid_tools.h [hunk 2]
```diff
@@ -130,6 +131,19 @@ namespace parallel
...
+     * boxes @p local_bboxes.
+     *
+     * This function was born to exchange bounding boxes describing the locally owned
```
Comment (davydden) [inline include/deal.II/distributed/grid_tools.h]: was born to -> is meant to

File: include/deal.II/distributed/grid_tools.h [hunk 3]
```diff
@@ -130,6 +131,19 @@ namespace parallel
...
+     * cells in a distributed triangulation obtained with the function
+     * GridTools::compute_mesh_predicate_bounding_box
```
Comment (davydden) [inline include/deal.II/distributed/grid_tools.h]: from the description i don't understand what it does. What do you mean by exchange? Is it govering all local boxes into a  single vector? You should probably clarify what is actually returned.

Comment (GivAlz) [reply include/deal.II/distributed/grid_tools.h]: Added a description, I hope it is good enough!

File: include/deal.II/distributed/grid_tools.h [hunk 4]
```diff
@@ -136,6 +137,22 @@ namespace parallel
...
 
     /**
+     * Exchange with all processors of the MPI communicator @p mpi_communicator the vector of bounding
+     * boxes @p local_bboxes.
+     *
+     * This function is meant to exchange bounding boxes describing the locally owned
+     * cells in a distributed triangulation obtained with the function
+     * GridTools::compute_mesh_predicate_bounding_box .
+     *
+     * The output vector' has length's size is number of processes of the MPI communicator:
```
Comment (tjhei) [inline include/deal.II/distributed/grid_tools.h]: please fix the grammar.

Comment (GivAlz) [reply include/deal.II/distributed/grid_tools.h]: Sorry...that's really painful to read even for me...

**Comment by GivAlz** [comment]
Ok, in better shape now grammar is! Thanks :)

**Comment by GivAlz** [comment]
Ok, build failed because missing #ifdef DEAL_II_WITH_MPI ... working on it!

File: source/distributed/grid_tools.cc [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            // Extracting the coordinates of each boundary point
+            loc_data_array[2*i*spacedim + d] = local_bboxes[i].get_boundary_points().first[d];
+            loc_data_array[2*i*spacedim + spacedim + d] = local_bboxes[i].get_boundary_points().second[d];
+          }
+
+      // Step 2: exchanging the size of local data
+      unsigned int n_procs = Utilities::MPI::n_mpi_processes(mpi_communicator);
+
+      // Dimension of the array to be exchanged (number of double)
+      int n_local_data = loc_data_array.size();
```
Comment (Rombur) [inline source/distributed/grid_tools.cc]: You could move this to line 36, set it right away to `n_local_data = 2*spacedim*n_bboxes`, and then use the value to initialize `loc_data_array`. You don't have to do it.

File: source/distributed/grid_tools.cc [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      // Vector to store the size of loc_data_array for every process
+      std::vector<int> size_all_data;
+      size_all_data.resize(n_procs);
```
Comment (Rombur) [inline source/distributed/grid_tools.cc]: `std::vector<int> size_all_data(n_procs);`

**Comment by GivAlz** [comment]
@Rombur Thank you for all the comments and the tips to have a cleaner, more elegant code. I tried to follow your suggestions for the #ifdef mpi...I hope it's ok now!

**Pull request merged**